### PR TITLE
Azure compare-and-exchange implementation

### PR DESCRIPTION
**Pull request by DaveCTurner**

### CODE REVIEW

**Comment by DaveCTurner** [comment]
NB I'm using Azure's synchronous client API here for the sake of simplicity/progress. I will open an issue to track the possibility of moving to the async API.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java [hunk 1]
```diff
@@ -795,4 +801,72 @@ private void consumeHttpRequestInfo() {
...
+                    .getBlobClient(blobPath);
+
+                if (blobClient.exists()) {
+                    final var leaseClient = new BlobLeaseClientBuilder().blobClient(blobClient).buildClient();
+                    final var leaseId = leaseClient.acquireLease(60);
+                    try {
+                        final var currentValue = getRegister(blobPath, containerPath, blobKey);
+                        if (currentValue.equals(OptionalLong.of(expected))) {
+                            blobClient.uploadWithResponse(
+                                new BlobParallelUploadOptions(BlobContainerUtils.getRegisterBlobContents(updated).streamInput())
```
Comment (fcofdez) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:839]: can we include the blob length here?

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java [hunk 2]
```diff
@@ -795,4 +801,72 @@ private void consumeHttpRequestInfo() {
...
+
+                final var blobClient = getAzureBlobServiceClientClient().getSyncClient()
+                    .getBlobContainerClient(container)
+                    .getBlobClient(blobPath);
+
+                if (blobClient.exists()) {
+                    final var leaseClient = new BlobLeaseClientBuilder().blobClient(blobClient).buildClient();
+                    final var leaseId = leaseClient.acquireLease(60);
+                    try {
+                        final var currentValue = getRegister(blobPath, containerPath, blobKey);
```
Comment (fcofdez) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:836]: I think we should get the register using the acquired lease? See https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob?tabs=azure-ad#request-headers

**Review by fcofdez — COMMENTED** [review]
I left a comment around reading the register value using the acquired lease.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java (hunk 2 above)
Comment (DaveCTurner) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:836]: ah yes you're right

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java (hunk 1 above)
Comment (DaveCTurner) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java]: yes, done in [7b28595](https://github.com/elastic/elasticsearch/pull/94276/commits/7b28595513b82eb8151858d0fd4c1b3ec3466acd)

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java (hunk 2 above)
Comment (DaveCTurner) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java]: done in [a03a836](https://github.com/elastic/elasticsearch/pull/94276/commits/a03a836f73c2736500cadc7ec13ebd27070feb73)

**Review by fcofdez — APPROVED** [review]
LGTM 👍

**Pull request merged**